include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -1235,13 +1300,105 @@ __parallel_find_or(oneapi::dpl::__internal::__device_backend_tag, _ExecutionPoli
         //The end of the scope  -  a point of synchronization (on temporary sycl buffer destruction)
     }
 
+    return __result;
+}
+
+// Base pattern for __parallel_or and __parallel_find. The execution depends on tag type _BrickTag.
+template <typename _ExecutionPolicy, typename _Brick, typename _BrickTag, typename... _Ranges>
+::std::conditional_t<
+    ::std::is_same_v<_BrickTag, __parallel_or_tag>, bool,
+    oneapi::dpl::__internal::__difference_t<typename oneapi::dpl::__ranges::__get_first_range_type<_Ranges...>::type>>
+__parallel_find_or(oneapi::dpl::__internal::__device_backend_tag, _ExecutionPolicy&& __exec, _Brick __f,
+                   _BrickTag __brick_tag, _Ranges&&... __rngs)
+{
+    using _CustomName = oneapi::dpl::__internal::__policy_kernel_name<_ExecutionPolicy>;
+    using _FindOrKernelOneWG =
+        oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_generator<__find_or_kernel_one_wg, _CustomName,
+                                                                               _Brick, _BrickTag, _Ranges...>;
+    using _FindOrKernel =
+        oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_generator<__find_or_kernel, _CustomName, _Brick,
+                                                                               _BrickTag, _Ranges...>;
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Calculate source data size
+
+    auto __rng_n = oneapi::dpl::__ranges::__get_first_range_size(__rngs...);
+    assert(__rng_n > 0);
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Calculate work-group size
+
+    // TODO: find a way to generalize getting of reliable work-group size
+    std::size_t __wgroup_size = oneapi::dpl::__internal::__max_work_group_size(__exec);
+#if _ONEDPL_COMPILE_KERNEL
+    auto __kernels = __internal::__kernel_compiler<_FindOrKernelOneWG, _FindOrKernel>::__compile(__exec);
+    auto __kernel_one_wg = __kernels[0];
+    auto __kernel_many_wg = __kernels[1];
+    __wgroup_size = std::min({__wgroup_size, oneapi::dpl::__internal::__kernel_work_group_size(__exec, __kernel_one_wg),
+                              oneapi::dpl::__internal::__kernel_work_group_size(__exec, __kernel_many_wg)});
+#endif
```

Comment:
Extending my comment about compiling the kernels: I would be interested to see if we even need it for the multi WG implementation since we don't use shared resources.